src/libraries/Common/src/Interop/Unix/System.Native/Interop.MemfdCreate.cs

```diff
@@ -0,0 +1,34 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.InteropServices;
+using System.Threading;
+using Microsoft.Win32.SafeHandles;
+
+internal static partial class Interop
+{
+    internal static partial class Sys
+    {
+        [LibraryImport(Libraries.SystemNative, EntryPoint = "SystemNative_MemfdCreate", StringMarshalling = StringMarshalling.Utf8, SetLastError = true)]
+        internal static partial SafeFileHandle MemfdCreate(string name);
+
+        [LibraryImport(Libraries.SystemNative, EntryPoint = "SystemNative_MemfdSupported", SetLastError = true)]
+        private static partial int MemfdSupportedImpl();
+
+        private static volatile int s_memfdSupported = -1;
+
+        internal static bool MemfdSupported
+        {
+            get
+            {
+                int result = Interlocked.CompareExchange(ref s_memfdSupported, -1, -1);
+                if (result == -1)
+                {
+                    result = MemfdSupportedImpl();
+                    Interlocked.Exchange(ref s_memfdSupported, result);
+                }
+                return result == 1;
```

Comment:
```suggestion
                int result = s_memfdSupported;
                if (result == -1)
                {
                    result = MemfdSupportedImpl();
                    s_memfdSupported = result;
                }
                return result == 1;
```

Is using `Interlocked` necessary here? Reading the 32-bit integer is atomic, and the field is `volatile`.

Reply:
Reading across P/Invoke boundary isn't atomic/thread-safe AFAIK.

Reply:
But a volatile read has acquire semantics. Does that not guarantee that the P/Invoke will not be executed out of order? Are P/Invokes even allowed to be executed out of order? (I don't know, just wondering)

Reply:
It is to make MemfdSupportedImpl() call only once per process and thread-safe. Solely relying on volatile does not prevent multiple calls to MemfdSupportedImpl() AFAIK. I am not an expert either, just following the existing pattern. Lets see what others think. 🙂

Reply:
> Reading across P/Invoke boundary isn't atomic/thread-safe

I am not sure what you mean.

I agree with @teo-tsirpanis that Interlocked.CompareExchange is unnecessary overhead to read the value.